File: src/main/scala/com/databricks/labs/hyperleaup/Creator.scala
package com.databricks.labs.hyperleaup

import java.io.File

import com.databricks.labs.hyperleaup.utils.SparkSessionWrapper
import com.tableau.hyperapi.{Catalog, Connection, CreateMode, HyperProcess, SchemaName, SqlType, TableDefinition, TableName, Telemetry}
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.types._
import com.tableau.hyperapi.Nullability.{NOT_NULLABLE, NULLABLE}
import java.nio.file.{Files, Path, Paths, StandardCopyOption}

import org.apache.log4j.Logger


class Creator(hyperFile: HyperFile) extends SparkSessionWrapper {

  private val logger: Logger = Logger.getLogger(this.getClass)

  private val _name: String = hyperFile.name
  private val _schema: String = hyperFile.schema
  private val _table: String = hyperFile.table
  private val _path: String = hyperFile.path
  private val _df: DataFrame = hyperFile.getDf

  /**
    * Converts a Spark StructField data type to corresponding Tableau sql type
    */
  private def convertStructField(column: StructField): TableDefinition.Column = {
    val name = column.name
    val dataType = column.dataType match {
      case StringType => SqlType.text()
      case IntegerType => SqlType.integer()
      case LongType => SqlType.bigInt()
      case DoubleType => SqlType.doublePrecision()
      case FloatType => SqlType.doublePrecision()
      case BooleanType => SqlType.bool()
      case DateType => SqlType.date()
      case TimestampType => SqlType.timestamp()
      case _ =>
        // check if the data type is a Decimal(precision,scale)
        if (column.dataType.toString.toLowerCase.startsWith("decimal")) {
          val decimalType = column.dataType.asInstanceOf[DecimalType]
          // Tableau SqlType.numeric has a maximum precision of 18
          val precision = if (decimalType.precision > 18) 18 else decimalType.precision
          val scale = decimalType.scale
          SqlType.numeric(precision, scale)
        } else {
          throw new IllegalArgumentException(s"Could not parse data type '${column.dataType}' in column : $name")
        }
    }
    val nullability = if (column.nullable) NULLABLE else NOT_NULLABLE
    new TableDefinition.Column(name, dataType, nullability)
  }

  /**
    * Returns a Tableau Table Definition given a StructType
    */
  private def getTableDefinition(schema: StructType): TableDefinition = {
    val columns = scala.collection.JavaConversions.seqAsJavaList(schema.map(convertStructField))
    new TableDefinition(new TableName(_schema, _table), columns)
  }

  /**
    * Checks if Databricks File System environment
    */
  private def isDbfsEnabled: Boolean = {
    spark.conf.get("spark.app.name").contains("Databricks") && Files.exists(Paths.get("/dbfs"))
  }

  /**
    * Writes a Spark DataFrame to a single CSV file on the local
    * filesystem.
    */
  private def writeCsvToLocalFilesystem(df: DataFrame): String = {
    // create a temp dir, prefixed with `hyperleaup`
    val tempDbfsDir: Path = Files.createTempDirectory("hyperleaup")
    val csvDbfsDir: String = tempDbfsDir.toString + "/csv"
    // write the DataFrame to local disk as a single CSV file
    df.coalesce(1).write.option("delimiter", ",").option("header", "true").mode("overwrite").csv(csvDbfsDir)
    // Spark DataFrameWriter will write metadata alongside the CSV,
    // ignore metedata and return only the CSV filename
    val sourceCsvDbfsFiles = new File(csvDbfsDir).listFiles.filter(_.isFile).filter(_.getName.endsWith("csv")).toList
    assert(sourceCsvDbfsFiles.length == 1) // should _always_ return length of 1
    sourceCsvDbfsFiles.head.toString
  }

  /**
    * Writes a Spark DataFrame to a single CSV file and moves the file
    * from the DBFS to a temp dir on the driver node, so that
    * the Tableau Hyper API can execute a COPY operation.
    * COPY operation is, by far, the most *efficient* method for creating a
    * Hyper file.
    */
  private def writeCsvToDbfs(df: DataFrame): String = {
    // create a temp dir, prefixed with `hyperleaup`
    val tempDbfsDir: Path = Files.createTempDirectory("hyperleaup")
    val csvDbfsDir: String = tempDbfsDir.toString + "/csv"
    // write the DataFrame to DBFS as a single CSV file
    df.coalesce(1).write.option("delimiter", ",").option("header", "true").mode("overwrite").csv(csvDbfsDir)
    // move the CSV file to a temp dir on the driver node
    moveCsvToDriverNode("/dbfs" + csvDbfsDir)
  }


  /**
    * Moves a CSV written to a Databricks Filesystem to temp
    * directory on the driver node.
    */
  private def moveCsvToDriverNode(dbfsPath: String): String = {
    // filter the directory contents at DBFS location to only CSV files
    val sourceCsvDbfsFiles = new File(dbfsPath).listFiles.filter(_.isFile).filter(_.getName.endsWith("csv")).toList
    assert(sourceCsvDbfsFiles.length == 1) // should _always_ return length of 1
    val sourceCsvDbfsPath: String = sourceCsvDbfsFiles.head.toString
    val destDriverNodePath: String = getDriverNodeTempDir(sourceCsvDbfsPath)
    // move CSV from DBFS location to temp dir on driver node
    Files.move(
      Paths.get(sourceCsvDbfsPath),
      Paths.get(destDriverNodePath),
      StandardCopyOption.REPLACE_EXISTING
    ).toString
  }

  /**
    * Generates a temp directory on the driver node taking the form:
    * "/tmp/hyperleaup${random_number}/csv/part-${uuid}.csv}"
    */
  private def getDriverNodeTempDir(sourcePath: String): String = {
    val parts = sourcePath.split("/")
    val len = parts.length
    val tempDirPath = s"/tmp/${parts(len-3)}/csv"
    val tempDir = new File(tempDirPath)
    tempDir.mkdirs()
    s"$tempDirPath/${parts(len-1)}"
  }

  /**
    * Creates a directory for the Hyper database given
    * a CSV location.
    * The Hyper database will be created next to `csv` dir.
    */
  def getHyperDatabasePath(targetCsvPath: String): String = {
    val parts = targetCsvPath.split("/")
    val len = parts.length
    val csvFilename = parts(len-1)
    // place Hyper database in `database` dir
    val destDatabasePath = targetCsvPath.replaceAll(s"csv/$csvFilename", "") + "database"
    val destDatabaseDir = new File(destDatabasePath)
    destDatabaseDir.mkdir()
    destDatabasePath
  }

  /**
    * Main function for writing a Spark DataFrame to a Tableau Hyper File on disk
    */
  def create(): String = {

    var hyperProcess: Option[HyperProcess] = None
    var connection: Option[Connection] = None

    try {

      hyperProcess = Some(new HyperProcess(Telemetry.DO_NOT_SEND_USAGE_DATA_TO_TABLEAU))

      // Check if Databricks Filesytem is present
      val csvPath = if (isDbfsEnabled) {
        logger.info("DBFS _IS_ enabled.")
        // A Hyper database cannot be opened using cloud storage,
        // so it must be created on the driver node
        writeCsvToDbfs(_df)
      } else {
        logger.info("DBFS is _NOT_ enabled.")
        writeCsvToLocalFilesystem(_df)
      }

      val hyperDatabasePath = getHyperDatabasePath(csvPath)
      val hyperDatabase: Path = Paths.get(s"$hyperDatabasePath/${_name}.hyper")

      connection = Some(new Connection(hyperProcess.get.getEndpoint, hyperDatabase.toString, CreateMode.CREATE_AND_REPLACE))

      val catalog: Catalog = connection.get.getCatalog
      catalog.createSchema(new SchemaName(_schema))
      val schema: StructType = _df.schema
      val tableDefinition: TableDefinition = getTableDefinition(schema)
      catalog.createTable(tableDefinition)

      // The most efficient method for adding data to a table is with the COPY command
      val copyCommand = s"COPY ${tableDefinition.getTableName} from '$csvPath' with (format csv, NULL 'NULL', delimiter ',', header)"
      val count = connection.get.executeCommand(copyCommand).getAsLong
      logger.info(s"Copied $count rows.")

      hyperDatabase.toString

    } catch {
      case e: Throwable => throw e
    } finally {
      connection.map(_.close)
      hyperProcess.map(_.close)
    }
  }
}

object Creator {
  def apply(hyperFile: HyperFile): Creator = new Creator(hyperFile)
}

File: hyperleaup/creator.py
import os
from shutil import copyfile
from typing import List, Any

from pyspark.sql import DataFrame
from pyspark.sql.types import *
from tableauhyperapi import SqlType, TableDefinition, NULLABLE, NOT_NULLABLE, TableName, HyperProcess, Telemetry, \
    CreateMode, Inserter, Connection
from pathlib import Path


def get_rows(df: DataFrame) -> List[Any]:
    """Returns an array of rows given a Spark DataFrame"""
    return df.rdd.map(lambda row: [x for x in row]).collect()


def convert_struct_field(column: StructField) -> TableDefinition.Column:
    """Converts a Spark StructField to a Tableau Hyper SqlType"""
    if column.dataType == IntegerType():
        sql_type = SqlType.int()
    elif column.dataType == LongType():
        sql_type = SqlType.big_int()
    elif column.dataType == DoubleType():
        sql_type = SqlType.double()
    elif column.dataType == FloatType():
        sql_type = SqlType.double()
    elif column.dataType == BooleanType():
        sql_type = SqlType.bool()
    elif column.dataType == DateType():
        sql_type = SqlType.date()
    elif column.dataType == TimestampType():
        sql_type = SqlType.timestamp()
    elif column.dataType == StringType():
        sql_type = SqlType.text()
    else:
        # Trap the DecimalType case
        if str(column.dataType).startswith("DecimalType"):
            # Max precision is only up to 18 decimal places in Tableau Hyper API
            precision = column.dataType.precision if column.dataType.precision <= 18 else 18
            scale = column.dataType.scale
            sql_type = SqlType.numeric(precision, scale)
        else:
            raise ValueError(f'Invalid StructField datatype for column `{column.name}` : {column.dataType}')
    nullable = NULLABLE if column.nullable else NOT_NULLABLE
    return TableDefinition.Column(name=column.name, type=sql_type, nullability=nullable)


def get_table_def(df: DataFrame, schema_name: str, table_name: str) -> TableDefinition:
    """Returns a Tableau TableDefintion given a Spark DataFrame"""
    schema = df.schema
    cols = list(map(convert_struct_field, schema))
    return TableDefinition(
        table_name=TableName("Extract", "Extract"),
        columns=cols
    )


def insert_data_into_hyper_file(data: List[Any], path: Path, table_def: TableDefinition):
    """Helper function that inserts data into a .hyper file."""
    with HyperProcess(telemetry=Telemetry.DO_NOT_SEND_USAGE_DATA_TO_TABLEAU) as hp:
        with Connection(endpoint=hp.endpoint,
                        database=path,
                        create_mode=CreateMode.CREATE_AND_REPLACE) as connection:
            connection.catalog.create_schema(schema=table_def.table_name.schema_name)
            connection.catalog.create_table(table_definition=table_def)
            with Inserter(connection, table_def) as inserter:
                inserter.add_rows(rows=data)
                inserter.execute()


def copy_data_into_hyper_file(csv_path: str, name: str, table_def: TableDefinition) -> str:
    """Helper function that copies data from a CSV file to a .hyper file."""
    hyper_database_path = f"/tmp/hyperleaup/{name}/{name}.hyper"
    with HyperProcess(telemetry=Telemetry.DO_NOT_SEND_USAGE_DATA_TO_TABLEAU) as hp:
        with Connection(endpoint=hp.endpoint,
                        database=Path(hyper_database_path),
                        create_mode=CreateMode.CREATE_AND_REPLACE) as connection:

            connection.catalog.create_schema(schema=table_def.table_name.schema_name)
            connection.catalog.create_table(table_definition=table_def)

            # The most efficient method for adding data to a table is with the COPY command
            copy_command = f"COPY {table_def.table_name} from '{csv_path}' with (format csv, NULL 'NULL', delimiter ',', header)"
            count = connection.execute_command(copy_command)
            print(f"Copied {count} rows.")

            return hyper_database_path


def write_csv_to_local_file_system(df: DataFrame, name: str) -> str:
    """Writes a Spark DataFrame to a single CSV file on the local filesystem."""
    tmp_dir = f"/tmp/hyperleaup/{name}/"

    # write the DataFrame to local disk as a single CSV file
    df.coalesce(1).write \
        .option("delimiter", ",") \
        .option("header", "true") \
        .option("inferSchema", "true")\
        .mode("overwrite").csv(tmp_dir)

    # Spark DataFrameWriter will write metadata alongside the CSV,
    # ignore metedata and return only the CSV filename
    for root_dir, dirs, files in os.walk(tmp_dir):
        for file in files:
            if file.endswith(".csv"):
                return f"{tmp_dir}/{file}"


def write_csv_to_dbfs(df: DataFrame, name: str) -> str:
    """Moves a CSV written to a Databricks Filesystem to a temp directory on the driver node."""
    tmp_dir = f"/tmp/hyperleaup/{name}/"

    # write the DataFrame to DBFS as a single CSV file
    df.coalesce(1).write \
        .option("delimiter", ",") \
        .option("header", "true") \
        .option("inferSchema", "true")\
        .mode("overwrite").csv(tmp_dir)

    # Spark DataFrameWriter will write metadata alongside the CSV,
    # ignore metedata and return only the CSV filename
    dbfs_tmp_dir = "/dbfs" + tmp_dir
    csv_file = None
    for root_dir, dirs, files in os.walk(dbfs_tmp_dir):
        for file in files:
            if file.endswith(".csv"):
                csv_file = file

    if csv_file is None:
        raise FileNotFoundError(f"CSV file '{tmp_dir}' not found on DBFS.")

    # Copy CSV from DBFS location to temp dir on driver node
    if not os.path.exists(tmp_dir):
        os.makedirs(tmp_dir)
    src_path = dbfs_tmp_dir + csv_file
    dest_path = tmp_dir + csv_file
    copyfile(src_path, dest_path)

    return dest_path


class Creator:

    def __init__(self, df: DataFrame, name: str, is_dbfs_enabled: bool = False):
        self.df = df
        self.name = name
        self.is_dbfs_enabled = is_dbfs_enabled

    def create(self) -> str:
        """Creates a Tableau Hyper File given a SQL statement"""
        # Write Spark DataFrame to CSV so that a file COPY can be done
        if not self.is_dbfs_enabled:
            print("Writing Spark DataFrame to local disk...")
            csv_path = write_csv_to_local_file_system(self.df, self.name)
        else:
            print("Writing Spark DataFrame to DBFS...")
            csv_path = write_csv_to_dbfs(self.df, self.name)

        # Convert the Spark DataFrame schema to a Tableau `TableDefinition`
        print("Generating Tableau Table Definition...")
        table_def = get_table_def(self.df, "Extract", "Extract")
        print("Done.")

        # COPY data into a Tableau .hyper file
        print("Copying data into Hyper File...")
        database_path = copy_data_into_hyper_file(csv_path, self.name, table_def)
        print("All done!")

        return database_path
